docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,276 @@
+
+
+```csharp
+string? TryGetMessage(string key);
+```
+
+The return value indicates success or failure, and carries the value if the value was found. In many cases, changing API signatures can improve how they communicate null values.
+
+However, for public libraries, or libraries with large user bases, you may prefer not introducing any API signature changes. For those cases, and other common patterns, you can apply attributes to more clearly define when an argument or return value may be `null`. Whether or not you consider changing the surface of your API, you'll likely find that type annotations alone aren't sufficient for describing when `null` values for arguments or return values. In those instances, you can apply attributes to more clearly describe an API. 
+
+## Attributes extend type annotations
+
+Several attributes have been added to express further information about the null state of variables. All code you wrote before C# 8 introduced nullable reference types was *null oblivious*. That means any reference type variable may be null, but null checks aren't required. Once your code is *nullable aware*, those rules change. Reference types should never be the `null` value, and nullable reference types must be checked against null before being dereferenced.
+
+The rules for your APIs are likely more complicated, as you saw with the `TryGetValue` API scenario. Many of your APIs have more complex rules for when variables can or can't be `null`. In these cases, you'll use one of these attributes to express those rules:
+
+- `AllowNull`: A non-nullable input argument may be null.
+- `DisallowNull`: A nullable input argument should never be null.
+- `MaybeNull`: A non-nullable return value may be null.
+- `NotNull`: A nullable return value will never be null.
+- `MaybeNullWhen`: A non-nullable `out` or `ref` argument may be null when the return value satisfies a condition.
+- `NotNullWhen`: A nullable `out` or `ref` argument may not be null when the return value satisfies a condition.
+- `NotNullIfNotNull`: a string return value isn't null when the input string argument isn't null.
+
+Adding these rules attributes gives the compiler more information about the rules for your API. The compiler will warn callers when they violate those rules. These attributes don't enable additional checks on your implementation.
+
+## Specify Preconditions: `AllowNull` and `DisallowNull`
+
+Consider a read / write property that never returns `null` because it has a reasonable default value. Callers pass `null` to the set accessor when set it to that default value. For example, consider a messaging system that asks for a screen name in a chat room. If none is provided, the system generates a random name:
```

Comment:
```suggestion
Consider a read / write property that never returns `null` because it has a reasonable default value. Callers pass `null` to the set accessor when setting it to that default value. For example, consider a messaging system that asks for a screen name in a chat room. If none is provided, the system generates a random name:
```